cf-net/cf-net.c

```diff
@@ -0,0 +1,766 @@
+// MAIN:
+//*******************************************************************
+
+int main(int argc, char **argv)
+{
+    CFNetOptions opts;
+    CFNetSetDefault(&opts);
+    char **args = NULL;
+    char *hostnames = NULL;
+    int ret = CFNetParse(argc, argv,                // Inputs
+                         &opts, &args, &hostnames); // Outputs
+    if (ret != 0)
+    {
+        exit(EXIT_FAILURE);
+    }
+    ret = CFNetRun(&opts, args, hostnames);  // Commands return exit code
+    return ret;
+}
+
+//*******************************************************************
+// INIT:
+//*******************************************************************
+
+static void CFNetSetDefault(CFNetOptions *opts){
+    assert(opts != NULL);
+    opts->debug       = false;
+    opts->verbose     = false;
+    opts->inform      = false;
+    opts->used_default= false;
+}
+
+static void CFNetInit()
+{
+    CryptoInitialize();
+    LoadSecretKeys();
+    cfnet_init(NULL, NULL);
+}
+
+//*******************************************************************
+// MAIN LOGIC:
+//*******************************************************************
+/*
+static void print_argv(char ** argv)
+{
+    for (int i = 0; argv[i] != NULL; ++i)
+    {
+        printf("argv[%i]: '%s'\n", i, argv[i]);
+    }
+}
+*/
+
+static char* RequireHostname(char *hostnames)
+{
+    if (hostnames == NULL)
+    {
+        char* policy_server = PolicyServerReadFile("/var/cfengine/");
+        if (policy_server == NULL)
+        {
+            printf("Error: no host name (and no policy_server.dat)");
+            exit(EXIT_FAILURE);
+        }
+        return policy_server;
+    }
+    return hostnames;
+}
+
+
+static int CFNetParse(int argc, char **argv,
+                      CFNetOptions *opts, char ***args, char **hostnames)
+{
+    assert(opts != NULL);
+    if (argc <= 1)
+    {
+        CFNetHelp(NULL);
+        return 0;
+    }
+    extern int optind;
+    extern char *optarg;
+    *hostnames = NULL;
+    int c = 0;
+    int start_index = 1;
+    const char *optstr = "+hMH:dvI"; // + means stop for non opt arg. :)
+    while ((c = getopt_long(argc, argv, optstr, OPTIONS, &start_index))
+            != -1)
+    {
+        switch (c)
+        {
+            case 'h':
+            {
+                CFNetHelp(NULL);
+                break;
+            }
+            case 'M':
+            {
+                // TODO: How do we actually add a man page
+                Writer *out = FileWriter(stdout);
+                ManPageWrite(out, "cf-net", time(NULL),
+                             CF_NET_SHORT_DESCRIPTION,
+                             CF_NET_MANPAGE_LONG_DESCRIPTION,
+                             OPTIONS, HINTS,
+                             true);
+                FileWriterDetach(out);
+                exit(EXIT_SUCCESS);
+                break;
+            }
+            case 'H':
+            {
+                *hostnames = xstrdup(optarg);
```

Comment:
I would consider either `free(hostnames)` or error message here if `-H` is given more than once.